packages/react-components/react-input/src/components/Input/useInputStyles.ts

```diff
@@ -243,16 +260,19 @@ export const useInputStyles_unstable = (state: InputState): InputState => {
   const invalid = `${state.input['aria-invalid']}` === 'true';
   const filled = appearance.startsWith('filled');
 
+  // Call exactly one of the two base className hooks. Each of these hooks is functionally identical, but with
+  // different styles applied, which makes it ok to conditionally change which hook is called.
+  const useRootClassName = disabled ? useRootNonInteractiveClassName : useRootInteractiveClassName;
```

Comment:
I borrowed this for the SpinButton `makeResetStyles` refactor but I'm wondering if it doesn't violate the spirit of the Rules of Hooks even though it doesn't trigger the lint error or React's runtime error?

The React doc examples all involve state and conditionally removing a call to a hook (i.e., calling 4 hook, then 3 conditionally) neither of which apply here but after thinking about this it _feels_ suspect despite it passing all the checks.

Reply:
I'd say it pushes the limits of the rules of hooks without technically breaking them. Internally each of the two `use*ClassName` hooks calls an identical set of other hooks to get the state for griffel. So ultimately they're both resulting in the same set of hooks being called when you swap them.

There are few possible ways to address this:
1. Leave it as-is in this PR, since it works.
2. Just call both hooks and pick which className gets used.
3. Use the `@griffel/core` version of makeStyles/makeResetStyles, which returns a non-hook function that takes some state as an argument. This would require having `@griffel/react` export a hook to get the state needed to pass to those functions.

```ts
import { makeResetStyles } from '@griffel/core';
import { useStyleState } from '@griffel/react'; // This doesn't exist, and the functions it would need aren't exported afaik

// These are 'get' functions, not 'use' hooks
const getRootInteractiveClassName = makeResetStyles(...};
const getRootNonInteractiveClassName = makeResetStyles(...};

// ...

const useStyles = () => {
  const styleState = useStyleState();

  const rootClassName = disabled ? getRootInteractiveClassName(styleState) : useRootInteractiveClassName(styleState);

}
```

Reply:
I'm going to merge this PR as-is, and make a new PR with a fourth option: Have a single useRootClassName that includes all of the "interactive" styles, and then update the disabled class to override any interactive styles that shouldn't apply.

Reply:
Follow-up PR:
* https://github.com/microsoft/fluentui/pull/26865

Reply:
I like that option. Will also copy that for SpinButton :)